rainier-core/src/main/scala/com/stripe/rainier/ir/Packer.scala

```diff
@@ -1,52 +1,64 @@
 package com.stripe.rainier.ir
 
+import scala.util.control.TailCalls
+import TailCalls.TailRec
+
 private class Packer(methodSizeLimit: Int) {
   private var methodDefs: List[MethodDef] = Nil
   def methods = methodDefs
 
   def pack(p: Expr): MethodRef = {
-    val (pExpr, _) = traverse(p, 0)
+    val (pExpr, _) = traverse(p, 0).result
     createMethod(UnaryIR(pExpr, NoOp))
   }
 
-  private def traverse(p: Expr, parentSize: Int): (Expr, Int) =
+  private def traverse(p: Expr, parentSize: Int): TailRec[(Expr, Int)] =
     p match {
       case v: VarDef => traverseVarDef(v, parentSize)
-      case _: Ref    => (p, 1)
+      case _: Ref    => TailCalls.done((p, 1))
     }
 
-  private def traverseVarDef(v: VarDef, parentSize: Int): (VarDef, Int) = {
-    val (ir, irSize) = traverseIR(v.rhs)
-    val (newIR, newSize) =
-      if ((irSize + parentSize) > methodSizeLimit)
-        (createMethod(ir), 1)
-      else
-        (ir, irSize)
-    (new VarDef(v.sym, newIR), newSize + 1)
-  }
+  private def traverseVarDef(v: VarDef,
+                             parentSize: Int): TailRec[(VarDef, Int)] =
+    TailCalls.tailcall(traverseIR(v.rhs).map {
+      case (ir, irSize) =>
+        val (newIR, newSize) =
+          if ((irSize + parentSize) > methodSizeLimit)
+            (createMethod(ir), 1)
+          else
+            (ir, irSize)
+        (new VarDef(v.sym, newIR), newSize + 1)
+    })
 
-  private def traverseIR(p: IR): (IR, Int) =
+  private def traverseIR(p: IR): TailRec[(IR, Int)] =
     p match {
       case b: BinaryIR =>
-        val (leftExpr, leftSize) =
-          traverse(b.left, 2)
-        val (rightExpr, rightSize) =
-          traverse(b.right, leftSize + 1)
-        (new BinaryIR(leftExpr, rightExpr, b.op), leftSize + rightSize + 1)
+        traverse(b.left, 2).flatMap {
```

Comment:
This `flatmap...map` could equally be

```
for {
  (leftExpr, leftSize) <- traverse(b.left, 2)
  (rightExpr, rightSize) <- traverse(b.right, leftSize + 1)
} yield (new BinaryIR(leftExpr, rightExpr, b.op), leftSize + rightSize + 1)
```

I was going to call it six of one, a half dozen of the other, but I think this way is a bit easier to parse.